backend/service/auditsink/slack/slack.go

```diff
@@ -104,3 +123,90 @@ func formatText(username string, event *auditv1.RequestEvent) string {
 
 	return messageText
 }
+
+// FormatCustomText parses out the audit event metdata request for the custom slack message text
+func FormatCustomText(message string, event *auditv1.RequestEvent) (string, error) {
+	tmpl, err := template.New("customText").Funcs(funcMap).Parse(message)
+	if err != nil {
+		return "", err
+	}
+
+	data, err := getAuditMetadata(event)
+	if err != nil {
+		return "", err
+	}
+
+	var buf bytes.Buffer
+	if err := tmpl.Execute(&buf, data); err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+// returns the API request/response details saved in an audit event
+func getAuditMetadata(event *auditv1.RequestEvent) (*auditEventMetadata, error) {
+	var requestMetadata map[string]interface{}
+	var responseMetadata map[string]interface{}
+
+	// proto -> json
+	reqJSON, err := protojson.Marshal(event.RequestMetadata.Body)
+	if err != nil {
+		return nil, err
+	}
+	respJSON, err := protojson.Marshal(event.ResponseMetadata.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// json -> map[string]interface{}
+	err = json.Unmarshal(reqJSON, &requestMetadata)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(respJSON, &responseMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	return &auditEventMetadata{
+		Request:  requestMetadata,
+		Response: responseMetadata,
+	}, nil
+}
+
+// for inputs that are type slice/map, returns a formatted slack list OR `N/A` if the slice/map is empty
+func slackList(data interface{}) (string, error) {
+	var text string
```

Comment:
this is a good use case for a string builder or buffer.
- https://www.calhoun.io/concatenating-and-building-strings-in-go/
- https://dave.cheney.net/high-performance-go-workshop/dotgo-paris.html#avoid_string_concatenation

Reply:
thanks for the links! went with `strings.Builder`